File: src/main/com/yubico/yubioath/model/YubiKeyNeo.java
/*
 * Copyright (c) 2013, Yubico AB.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *  Redistributions of source code must retain the above copyright
 *   notice, this list of conditions and the following disclaimer.
 *
 *  Redistributions in binary form must reproduce the above copyright
 *   notice, this list of conditions and the following
 *   disclaimer in the documentation and/or other materials provided
 *   with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
 * CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
 * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
 * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
 * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
 * TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
 * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

package com.yubico.yubioath.model;

import android.nfc.tech.IsoDep;
import android.util.Log;
import com.yubico.yubioath.exc.*;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.io.IOException;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.util.*;

/**
 * Created with IntelliJ IDEA.
 * User: dain
 * Date: 8/23/13
 * Time: 3:57 PM
 * To change this template use File | Settings | File Templates.
 */
public class YubiKeyNeo {
    private static final byte[] APDU_OK = {(byte) 0x90, 0x00};
    private static final byte[] APDU_FILE_FULL = {(byte) 0x6a, (byte) 0x84};

    public static final byte NAME_TAG = 0x71;
    public static final byte NAME_LIST_TAG = 0x72;
    public static final byte KEY_TAG = 0x73;
    public static final byte CHALLENGE_TAG = 0x74;
    public static final byte RESPONSE_TAG = 0x75;
    public static final byte T_RESPONSE_TAG = 0x76;
    public static final byte NO_RESPONSE_TAG = 0x77;
    public static final byte PROPERTY_TAG = 0x78;
    public static final byte VERSION_TAG = 0x79;
    public static final byte IMF_TAG = 0x7a;

    public static final byte PUT_INS = 0x01;
    public static final byte DELETE_INS = 0x02;
    public static final byte SET_CODE_INS = 0x03;
    public static final byte RESET_INS = 0x04;

    public static final byte LIST_INS = (byte) 0xa1;
    public static final byte CALCULATE_INS = (byte) 0xa2;
    public static final byte VALIDATE_INS = (byte) 0xa3;
    public static final byte CALCULATE_ALL_INS = (byte) 0xa4;
    public static final byte SEND_REMAINING_INS = (byte) 0xa5;

    public static final byte HMAC_MASK = 0x0f;
    public static final byte HMAC_SHA1 = 0x01;
    public static final byte HMAC_SHA256 = 0x02;

    public static final byte OATH_MASK = (byte) 0xf0;
    public static final byte HOTP_TYPE = 0x10;
    public static final byte TOTP_TYPE = 0x20;


    //APDU CL INS P1 P2 L ...
    //DATA 00  00 00 00 00 ...
    private static final byte[] SELECT_COMMAND = {0x00, (byte) 0xa4, 0x04, 0x00, 0x08, (byte) 0xa0, 0x00, 0x00, 0x05, 0x27, 0x21, 0x01, 0x01};
    private static final byte[] CALCULATE_ALL_COMMAND = {0x00, CALCULATE_ALL_INS, 0x00, 0x01, 0x0a, CHALLENGE_TAG, 0x08, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};
    private static final byte[] CALCULATE_COMMAND = {0x00, CALCULATE_INS, 0x00, 0x01, 0x00};
    private static final byte[] SET_LOCK_COMMAND = {0x00, SET_CODE_INS, 0x00, 0x00, 0x00};
    private static final byte[] UNLOCK_COMMAND = {0x00, VALIDATE_INS, 0x00, 0x00, 0x00};
    private static final byte[] PUT_COMMAND = {0x00, PUT_INS, 0x00, 0x00, 0x00};
    private static final byte[] DELETE_COMMAND = {0x00, DELETE_INS, 0x00, 0x00, 0x00};
    private static final byte[] SEND_REMAINING_COMMAND = {0x00, SEND_REMAINING_INS, 0x00, 0x00, 0x00};

    private static final int[] MOD = {1, 10, 100, 1000, 10000, 100000, 1000000, 10000000, 100000000};


    private final KeyManager keyManager;
    private final IsoDep isoTag;
    private final byte[] id;

    public YubiKeyNeo(KeyManager keyManager, IsoDep isoTag) throws IOException, AppletSelectException {
        this.keyManager = keyManager;
        this.isoTag = isoTag;

        isoTag.connect();
        isoTag.setTimeout(3000);
        byte[] resp = isoTag.transceive(SELECT_COMMAND);
        if (!compareStatus(resp, APDU_OK)) {
            throw new AppletMissingException();
        }

        int offset = 0;
        byte[] version = parseBlock(resp, offset, VERSION_TAG);
        offset += version.length + 2;

        checkVersion(version);

        id = parseBlock(resp, offset, NAME_TAG);
        offset += id.length + 2;
        if (resp.length - offset - 4 > 0) {
            byte[] challenge = parseBlock(resp, offset, CHALLENGE_TAG);
            unlock(challenge);
        }
    }

    public byte[] getId() {
        return id;
    }

    public String getDisplayName(String defaultName) {
        return keyManager.getDisplayName(id, defaultName);
    }

    public void setDisplayName(String name) {
        keyManager.setDisplayName(id, name);
    }

    private void unlock(byte[] challenge) throws IOException, PasswordRequiredException {
        byte[] secret = keyManager.getSecret(id);
        byte[] altSecret = keyManager.getAltSecret(id);
        if (secret == null && altSecret == null) {
            throw new PasswordRequiredException("Password is missing!", id, true);
        }

        if (secret != null && doUnlock(challenge, secret)) { //Try the main password
            keyManager.storeAltSecret(id, new byte[0], true);
            return;
        } else if (altSecret != null && doUnlock(challenge, altSecret)) { //Try a password that might have been set
            keyManager.promoteAltSecret(id);
            return;
        }

        throw new PasswordRequiredException("Password is incorrect!", id, false);
    }

    private boolean doUnlock(byte[] challenge, byte[] secret) throws IOException {
        byte[] response = hmacSha1(secret, challenge);
        byte[] myChallenge = new byte[8];
        SecureRandom random = new SecureRandom();
        random.nextBytes(myChallenge);
        byte[] myResponse = hmacSha1(secret, myChallenge);

        byte[] data = new byte[UNLOCK_COMMAND.length + 2 + response.length + 2 + myChallenge.length];
        System.arraycopy(UNLOCK_COMMAND, 0, data, 0, UNLOCK_COMMAND.length);
        int offset = 4;
        data[offset++] = (byte) (data.length - 5);

        data[offset++] = RESPONSE_TAG;
        data[offset++] = (byte) response.length;
        System.arraycopy(response, 0, data, offset, response.length);
        offset += response.length;

        data[offset++] = CHALLENGE_TAG;
        data[offset++] = (byte) myChallenge.length;
        System.arraycopy(myChallenge, 0, data, offset, myChallenge.length);
        offset += myChallenge.length;

        byte[] resp = isoTag.transceive(data);

        if (compareStatus(resp, APDU_OK)) {
            byte[] neoResponse = parseBlock(resp, 0, RESPONSE_TAG);
            if (Arrays.equals(myResponse, neoResponse)) {
                return true;
            }
        }

        return false;
    }

    private void unsetLockCode() throws IOException {
        keyManager.storeAltSecret(id, new byte[0], true);
        byte[] data = new byte[SET_LOCK_COMMAND.length + 2];
        System.arraycopy(SET_LOCK_COMMAND, 0, data, 0, SET_LOCK_COMMAND.length);
        data[4] = 2;
        data[5] = KEY_TAG;

        requireStatus(isoTag.transceive(data), APDU_OK);
        keyManager.promoteAltSecret(id);
    }

    public void setLockCode(String code, boolean remember) throws IOException {
        byte[] secret = KeyManager.calculateSecret(code, id);
        if (secret.length == 0) {
            unsetLockCode();
            return;
        } else if(Arrays.equals(secret, keyManager.getSecret(id))) {
            return;
        }

        keyManager.storeAltSecret(id, secret, remember); //Store but don't overwrite. If we get a failure below we won't know if the password was set or not.

        byte[] challenge = new byte[8];
        SecureRandom random = new SecureRandom();
        random.nextBytes(challenge);
        byte[] response = hmacSha1(secret, challenge);

        byte[] data = new byte[SET_LOCK_COMMAND.length + 3 + secret.length + 2 + challenge.length + 2 + response.length];
        System.arraycopy(SET_LOCK_COMMAND, 0, data, 0, SET_LOCK_COMMAND.length);
        int offset = 4;
        data[offset++] = (byte) (data.length - 5);

        data[offset++] = KEY_TAG;
        data[offset++] = (byte) (secret.length + 1);
        data[offset++] = TOTP_TYPE | HMAC_SHA1;
        System.arraycopy(secret, 0, data, offset, secret.length);
        offset += secret.length;

        data[offset++] = CHALLENGE_TAG;
        data[offset++] = (byte) challenge.length;
        System.arraycopy(challenge, 0, data, offset, challenge.length);
        offset += challenge.length;

        data[offset++] = RESPONSE_TAG;
        data[offset++] = (byte) response.length;
        System.arraycopy(response, 0, data, offset, response.length);

        requireStatus(isoTag.transceive(data), APDU_OK);
        keyManager.promoteAltSecret(id);
    }

    public void storeCode(String name, byte[] key, byte type, int digits, int counter) throws IOException {
        byte[] nameBytes = name.getBytes();
        byte[] counterBytes = null;
        int length = PUT_COMMAND.length + 2 + nameBytes.length + 4 + key.length;
        if (counter > 0) {
            length += 6;
            counterBytes = new byte[4];
            counterBytes[0] = (byte) (counter >>> 24);
            counterBytes[1] = (byte) (counter >>> 16);
            counterBytes[2] = (byte) (counter >>> 8);
            counterBytes[3] = (byte) counter;
        }
        byte[] data = new byte[length];
        System.arraycopy(PUT_COMMAND, 0, data, 0, PUT_COMMAND.length);
        int offset = 4;
        data[offset++] = (byte) (data.length - 5);

        data[offset++] = NAME_TAG;
        data[offset++] = (byte) nameBytes.length;
        System.arraycopy(nameBytes, 0, data, offset, nameBytes.length);
        offset += nameBytes.length;

        data[offset++] = KEY_TAG;
        data[offset++] = (byte) (key.length + 2);
        data[offset++] = type;
        data[offset++] = (byte) digits;
        System.arraycopy(key, 0, data, offset, key.length);
        offset += key.length;

        if (counterBytes != null) {
            data[offset++] = IMF_TAG;
            data[offset++] = (byte) counterBytes.length;
            System.arraycopy(counterBytes, 0, data, offset, counterBytes.length);
            offset += counterBytes.length;
        }

        byte[] resp = isoTag.transceive(data);
        if (compareStatus(resp, APDU_FILE_FULL)) {
            throw new StorageFullException("No more room for OATH credentials!");
        } else {
            requireStatus(resp, APDU_OK);
        }
    }

    public void deleteCode(String name) throws IOException {
        byte[] nameBytes = name.getBytes();
        byte[] data = new byte[DELETE_COMMAND.length + 2 + nameBytes.length + 2];
        System.arraycopy(DELETE_COMMAND, 0, data, 0, DELETE_COMMAND.length);
        int offset = 4;
        data[offset++] = (byte) (data.length - 5);

        data[offset++] = NAME_TAG;
        data[offset++] = (byte) nameBytes.length;
        System.arraycopy(nameBytes, 0, data, offset, nameBytes.length);
        requireStatus(isoTag.transceive(data), APDU_OK);
    }

    public String readHotpCode(String name) throws IOException {
        byte[] nameBytes = name.getBytes();
        byte[] data = new byte[CALCULATE_COMMAND.length + 2 + nameBytes.length + 2];
        System.arraycopy(CALCULATE_COMMAND, 0, data, 0, CALCULATE_COMMAND.length);
        int offset = 4;
        data[offset++] = (byte) (data.length - 5);

        data[offset++] = NAME_TAG;
        data[offset++] = (byte) nameBytes.length;
        System.arraycopy(nameBytes, 0, data, offset, nameBytes.length);
        offset += nameBytes.length;

        data[offset++] = CHALLENGE_TAG;
        data[offset++] = 0;

        byte[] resp = requireStatus(send(data), APDU_OK);
        return codeFromTruncated(parseBlock(resp, 0, T_RESPONSE_TAG));
    }

    public List<Map<String, String>> getCodes(long timestamp) throws IOException {
        List<Map<String, String>> codes = new ArrayList<Map<String, String>>();

        byte[] command = new byte[CALCULATE_ALL_COMMAND.length];
        System.arraycopy(CALCULATE_ALL_COMMAND, 0, command, 0, CALCULATE_ALL_COMMAND.length);
        int offset = CALCULATE_ALL_COMMAND.length - 4;
        command[offset++] = (byte) (timestamp >> 24);
        command[offset++] = (byte) (timestamp >> 16);
        command[offset++] = (byte) (timestamp >> 8);
        command[offset++] = (byte) timestamp;

        byte[] resp = requireStatus(send(command), APDU_OK);

        offset = 0;
        while (resp[offset] == NAME_TAG) {
            byte[] name = parseBlock(resp, offset, NAME_TAG);
            offset += name.length + 2;
            byte responseType = resp[offset];
            byte[] hashBytes = parseBlock(resp, offset, responseType);
            offset += hashBytes.length + 2;

            Map<String, String> oathCode = new HashMap<String, String>();
            oathCode.put("label", new String(name));
            switch (responseType) {
                case T_RESPONSE_TAG:
                    oathCode.put("code", codeFromTruncated(hashBytes));
                    break;
                case NO_RESPONSE_TAG:
                    oathCode.put("code", null);
                    break;
                default:
                    oathCode.put("code", "<invalid code>");
            }
            Log.d("yubioath", "label: " + oathCode.get("label"));
            codes.add(oathCode);
        }

        return codes;
    }

    private byte[] send(byte[] command) throws IOException {
        byte[] resp = isoTag.transceive(command);
        byte[] buf = new byte[2048];
        int offset = 0;

        while (resp[resp.length - 2] == 0x61) {
            System.arraycopy(resp, 0, buf, offset, resp.length - 2);
            offset += resp.length - 2;
            resp = isoTag.transceive(SEND_REMAINING_COMMAND);
        }

        System.arraycopy(resp, 0, buf, offset, resp.length);
        byte[] properlySized = new byte[offset + resp.length];
        System.arraycopy(buf, 0, properlySized, 0, properlySized.length);

        return properlySized;
    }

    public void close() throws IOException {
        isoTag.close();
    }

    private static void checkVersion(byte[] version) throws UnsupportedAppletException {
        byte major = version[0];

        if (major > 0) {
            throw new UnsupportedAppletException(version);
        }
    }

    private static boolean compareStatus(byte[] apdu, byte[] status) {
        return apdu[apdu.length - 2] == status[0] && apdu[apdu.length - 1] == status[1];
    }

    private static byte[] requireStatus(byte[] apdu, byte[] status) throws IOException {
        if (!compareStatus(apdu, status)) {
            String expected = String.format("%02x%02x", 0xff & status[0], 0xff & status[1]).toUpperCase();
            String actual = String.format("%02x%02x", 0xff & apdu[apdu.length - 2], 0xff & apdu[apdu.length - 1]).toUpperCase();
            throw new IOException("Require APDU status: " + expected + ", got " + actual);
        }
        return apdu;
    }

    private static byte[] parseBlock(byte[] data, int offset, byte identifier) throws IOException {
        if (data[offset] == identifier) {
            int length = data[offset + 1];
            byte[] block = new byte[length];
            System.arraycopy(data, offset + 2, block, 0, length);
            return block;
        } else {
            throw new IOException("Require block type: " + identifier + ", got: " + data[offset]);
        }
    }

    private static byte[] hmacSha1(byte[] key, byte[] data) {
        try {
            Mac mac = Mac.getInstance("HmacSHA1");
            SecretKeySpec secret = new SecretKeySpec(key, mac.getAlgorithm());
            mac.init(secret);
            return mac.doFinal(data);
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        } catch (InvalidKeyException e) {
            throw new RuntimeException(e);
        }
    }

    private static String codeFromTruncated(byte[] data) {
        int num_digits = data[0];
        int code = (data[1] << 24) | ((data[2] & 0xff) << 16) | ((data[3] & 0xff) << 8) | (data[4] & 0xff);
        return String.format("%0" + num_digits + "d", code % MOD[num_digits]);
    }
}

File: app/src/main/java/com/yubico/yubioath/model/YubiKeyNeo.kt
/*
 * Copyright (c) 2013, Yubico AB.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *  Redistributions of source code must retain the above copyright
 *   notice, this list of conditions and the following disclaimer.
 *
 *  Redistributions in binary form must reproduce the above copyright
 *   notice, this list of conditions and the following
 *   disclaimer in the documentation and/or other materials provided
 *   with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
 * CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
 * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
 * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
 * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
 * TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
 * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

package com.yubico.yubioath.model

import com.yubico.yubioath.exc.*

import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec
import java.io.IOException
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException
import java.security.SecureRandom
import java.util.*
import nordpol.IsoCard
import java.io.ByteArrayOutputStream
import java.io.Closeable

/**
 * Created with IntelliJ IDEA.
 * User: dain
 * Date: 8/23/13
 * Time: 3:57 PM
 * To change this template use File | Settings | File Templates.
 */

class YubiKeyNeo @Throws(IOException::class, AppletSelectException::class)
constructor(private val keyManager: KeyManager, private val isoTag: IsoCard) : Closeable {
    val id: ByteArray
    private var challenge: ByteArray = ByteArray(0)

    init {
        isoTag.connect()
        isoTag.timeout = 3000
        val resp = isoTag.transceive(SELECT_COMMAND)
        if (!compareStatus(resp, APDU_OK)) {
            throw AppletMissingException()
        }

        var offset = 0
        val version = parseBlock(resp, offset, VERSION_TAG)
        offset += version.size + 2

        checkVersion(version)

        id = parseBlock(resp, offset, NAME_TAG)
        offset += id.size + 2
        if (resp.size - offset - 4 > 0) {
            challenge = parseBlock(resp, offset, CHALLENGE_TAG)
        }
    }

    fun isLocked(): Boolean = challenge.isNotEmpty()

    @Throws(IOException::class, PasswordRequiredException::class)
    fun unlock() {
        val secrets = keyManager.getSecrets(id)

        if (secrets.isEmpty()) {
            throw PasswordRequiredException("Password is missing!", id, true)
        }

        for (secret in secrets) {
            if (doUnlock(challenge, secret)) {
                keyManager.setOnlySecret(id, secret)
                challenge = ByteArray(0)
                return
            }
        }

        throw PasswordRequiredException("Password is incorrect!", id, false)
    }

    @Throws(IOException::class)
    private fun doUnlock(challenge: ByteArray, secret: ByteArray): Boolean {
        val response = hmacSha1(secret, challenge)
        val myChallenge = ByteArray(8)
        val random = SecureRandom()
        random.nextBytes(myChallenge)
        val myResponse = hmacSha1(secret, myChallenge)

        val data = apdu(VALIDATE_INS) {
            tlv(RESPONSE_TAG, response)
            tlv(CHALLENGE_TAG, myChallenge)
        }

        val resp = send(data)

        if (compareStatus(resp, APDU_OK)) {
            return Arrays.equals(myResponse, parseBlock(resp, 0, RESPONSE_TAG))
        }

        return false
    }

    @Throws(IOException::class)
    private fun unsetLockCode() {
        val data = apdu(SET_CODE_INS) {
            tlv(KEY_TAG, byteArrayOf())
        }

        requireStatus(send(data), APDU_OK)
        keyManager.storeSecret(id, ByteArray(0), true)
    }

    @Throws(IOException::class)
    fun setLockCode(code: String, remember: Boolean) {
        val secret = KeyManager.calculateSecret(code, id, false)
        if (secret.size == 0) {
            unsetLockCode()
            return
        } else if (keyManager.getSecrets(id).contains(secret)) {
            return
        }

        val challenge = ByteArray(8)
        val random = SecureRandom()
        random.nextBytes(challenge)
        val response = hmacSha1(secret, challenge)

        val data = apdu(SET_CODE_INS) {
            tlv(KEY_TAG, byteArrayOf((TOTP_TYPE.toInt() or HMAC_SHA1.toInt()).toByte()) + secret)
            tlv(CHALLENGE_TAG, challenge)
            tlv(RESPONSE_TAG, response)
        }

        requireStatus(send(data), APDU_OK)
        keyManager.setOnlySecret(id, secret)
    }

    @Throws(IOException::class)
    fun storeCode(name: String, key: ByteArray, type: Byte, digits: Int, counter: Int) {
        val data = apdu(PUT_INS) {
            tlv(NAME_TAG, name.toByteArray())
            tlv(KEY_TAG, byteArrayOf(type, digits.toByte()) + if(key.size > 64) sha1(key) else key)
            if(counter >= 0) add(IMF_TAG, 4,
                    counter.ushr(24),
                    counter.ushr(16),
                    counter.ushr(8),
                    counter)
        }

        val resp = send(data)
        if (compareStatus(resp, APDU_FILE_FULL)) {
            throw StorageFullException("No more room for OATH credentials!")
        } else {
            requireStatus(resp, APDU_OK)
        }
    }

    @Throws(IOException::class)
    fun deleteCode(name: String) {
        val data = apdu(DELETE_INS) {
            tlv(NAME_TAG, name.toByteArray())
        }
        requireStatus(send(data), APDU_OK)
    }

    @Throws(IOException::class)
    fun readHotpCode(name: String): String {
        val data = apdu(CALCULATE_INS, p2=1) {
            tlv(NAME_TAG, name.toByteArray())
            tlv(CHALLENGE_TAG, byteArrayOf())
        }
        val resp = requireStatus(send(data), APDU_OK)
        return codeFromTruncated(parseBlock(resp, 0, T_RESPONSE_TAG))
    }

    @Throws(IOException::class)
    fun getCodes(timestamp: Long): List<Map<String, String>> {
        val codes = ArrayList<Map<String, String>>()

        val data = apdu(CALCULATE_ALL_INS, p2=1) {
            add(CHALLENGE_TAG, 4,
                    timestamp.shr(24),
                    timestamp.shr(16),
                    timestamp.shr(8),
                    timestamp
            )
        }
        val resp = requireStatus(send(data), APDU_OK)

        var offset = 0
        while (resp[offset] == NAME_TAG) {
            val name = parseBlock(resp, offset, NAME_TAG)
            offset += name.size + 2
            val responseType = resp[offset]
            val hashBytes = parseBlock(resp, offset, responseType)
            offset += hashBytes.size + 2

            val oathCode = HashMap<String, String>()
            val credentialName = String(name)
            oathCode.put("label", credentialName)
            when (responseType) {
                T_RESPONSE_TAG -> {
                    val code: String
                    if (credentialName.startsWith("Steam:")) {
                        code = steamCodeFromTruncated(hashBytes)
                    } else {
                        code = codeFromTruncated(hashBytes)
                    }
                    oathCode.put("code", code)
                }
                NO_RESPONSE_TAG -> oathCode.put("code", "")
                else -> oathCode.put("code", "<invalid code>")
            }
            codes.add(oathCode)
        }

        return codes
    }

    @Throws(IOException::class)
    private fun send(command: ByteArray): ByteArray {
        var resp = isoTag.transceive(command)
        val buf = ByteArrayOutputStream(2048)

        while (resp[resp.size - 2] == APDU_DATA_REMAINING_SW1) {
            buf.write(resp, 0, resp.size - 2)
            resp = isoTag.transceive(SEND_REMAINING_COMMAND)
        }

        buf.write(resp)
        return buf.toByteArray()
    }

    @Throws(IOException::class)
    override fun close() {
        isoTag.close()
    }

    companion object {
        private val APDU_OK = byteArrayOf(0x90.toByte(), 0x00)
        private val APDU_FILE_FULL = byteArrayOf(0x6a.toByte(), 0x84.toByte())
        const private val APDU_DATA_REMAINING_SW1 = 0x61.toByte()

        const private val NAME_TAG: Byte = 0x71
        const private val NAME_LIST_TAG: Byte = 0x72
        const private val KEY_TAG: Byte = 0x73
        const private val CHALLENGE_TAG: Byte = 0x74
        const private val RESPONSE_TAG: Byte = 0x75
        const private val T_RESPONSE_TAG: Byte = 0x76
        const private val NO_RESPONSE_TAG: Byte = 0x77
        const private val PROPERTY_TAG: Byte = 0x78
        const private val VERSION_TAG: Byte = 0x79
        const private val IMF_TAG: Byte = 0x7a

        const private val PUT_INS: Byte = 0x01
        const private val DELETE_INS: Byte = 0x02
        const private val SET_CODE_INS: Byte = 0x03
        const private val RESET_INS: Byte = 0x04

        const private val LIST_INS = 0xa1.toByte()
        const private val CALCULATE_INS = 0xa2.toByte()
        const private val VALIDATE_INS = 0xa3.toByte()
        const private val CALCULATE_ALL_INS = 0xa4.toByte()
        const private val SEND_REMAINING_INS = 0xa5.toByte()

        const val HMAC_MASK: Byte = 0x0f
        const val HMAC_SHA1: Byte = 0x01
        const val HMAC_SHA256: Byte = 0x02

        const val OATH_MASK = 0xf0.toByte()
        const val HOTP_TYPE: Byte = 0x10
        const val TOTP_TYPE: Byte = 0x20


        //APDU CL INS P1 P2 L ...
        //DATA 00  00 00 00 00 ...
        private val SELECT_COMMAND = byteArrayOf(0x00, 0xa4.toByte(), 0x04, 0x00, 0x08, 0xa0.toByte(), 0x00, 0x00, 0x05, 0x27, 0x21, 0x01, 0x01)
        private val SEND_REMAINING_COMMAND = byteArrayOf(0x00, SEND_REMAINING_INS, 0x00, 0x00, 0x00)

        private val MOD = intArrayOf(1, 10, 100, 1000, 10000, 100000, 1000000, 10000000, 100000000)

        @Throws(UnsupportedAppletException::class)
        private fun checkVersion(version: ByteArray) {
            // All versions currently
        }

        private fun compareStatus(apdu: ByteArray, status: ByteArray): Boolean {
            return apdu[apdu.size - 2] == status[0] && apdu[apdu.size - 1] == status[1]
        }

        @Throws(IOException::class)
        private fun requireStatus(apdu: ByteArray, status: ByteArray): ByteArray {
            if (!compareStatus(apdu, status)) {
                val expected = "%02x%02x".format(0xff and status[0].toInt(), 0xff and status[1].toInt()).toUpperCase()
                val actual = "%02x%02x".format(0xff and apdu[apdu.size - 2].toInt(), 0xff and apdu[apdu.size - 1].toInt()).toUpperCase()
                throw IOException("Require APDU status: $expected, got $actual")
            }
            return apdu
        }

        @Throws(IOException::class)
        private fun parseBlock(data: ByteArray, offset: Int, identifier: Byte): ByteArray {
            if (data[offset] == identifier) {
                val length = data[offset + 1].toInt()
                val block = ByteArray(length)
                System.arraycopy(data, offset + 2, block, 0, length)
                return block
            } else {
                throw IOException("Require block type: " + identifier + ", got: " + data[offset])
            }
        }

        private fun sha1(data: ByteArray): ByteArray {
            try {
                val md = MessageDigest.getInstance("SHA1")
                return md.digest(data)
            } catch (e: NoSuchAlgorithmException) {
                throw RuntimeException(e)
            }

        }

        private fun hmacSha1(key: ByteArray, data: ByteArray): ByteArray {
            val mac = Mac.getInstance("HmacSHA1")
            val secret = SecretKeySpec(key, mac.algorithm)
            mac.init(secret)
            return mac.doFinal(data)
        }

        private fun codeFromTruncated(data: ByteArray): String {
            val intData = data.map { it.toInt() }.toIntArray()  //Bitwise operations require ints.
            val num_digits = intData[0]
            val code = (intData[1] shl 24) or ((intData[2] and 0xff) shl 16) or ((intData[3] and 0xff) shl 8) or (intData[4] and 0xff)
            return String.format("%0" + num_digits + "d", code % MOD[num_digits])
        }

        private val STEAM_CHARS = "23456789BCDFGHJKMNPQRTVWXY"

        private fun steamCodeFromTruncated(data: ByteArray): String {
            val intData = data.map { it.toInt() }.toIntArray()  //Bitwise operations require ints.
            var code = (intData[1] shl 24) or ((intData[2] and 0xff) shl 16) or ((intData[3] and 0xff) shl 8) or (intData[4] and 0xff)
            val buf = StringBuilder()
            for (i in 0..4) {
                buf.append(STEAM_CHARS[code % STEAM_CHARS.length])
                code /= STEAM_CHARS.length
            }
            return buf.toString()
        }

        private fun ByteArrayOutputStream.tlv(tag: Byte, data: ByteArray) {
            write(tag.toInt())
            write(data.size)
            write(data)
        }
        private fun ByteArrayOutputStream.add(vararg values: Number) {
            for(v in values) write(v.toInt())
        }
        private fun apdu(ins:Byte, p1: Byte = 0, p2: Byte = 0, func: ByteArrayOutputStream.() -> Unit): ByteArray {
            return ByteArrayOutputStream().apply {
                add(0, ins, p1, p2, 0)
                func()
            }.toByteArray().apply {
                set(4, (size - 5).toByte())  // Fix length field
            }
        }
    }
}
